modules/message-lane/src/lib.rs

```diff
@@ -90,19 +159,237 @@ decl_module! {
+				match lane.receive_message(message.key.nonce, message.data, &mut dispatcher) {
+					ReceiveMessageResult::Invalid => invalid_messages += 1,
+					ReceiveMessageResult::Processed => processed_messages += 1,
+					ReceiveMessageResult::Queued => queued_messages += 1,
+				}
+			}
+
+			let weight_left = dispatcher.weight_left();
+			frame_support::debug::trace!(
+				target: "runtime",
+				"Received messages: total={}, invalid={}, processed={}, queued={}. Weight spent: {}, left: {}",
+				total_messages,
+				invalid_messages,
+				processed_messages,
+				queued_messages,
+				max_dispatch_weight.saturating_sub(weight_left),
+				weight_left,
+			);
+
+			// ok - we have processed or queued all bundled messages. But what if relayer acts aggressivly
+			// (i.e. `max_dispatch_weight` is larger than `messages_dispatch_weight`), or if some messages
+			// have been queued? Then we may have some capacity left from `max_dispatch_weight`.
+			// => let's not waste it and try to process queued messages
+			let mut queued_dispatcher = T::QueuedMessageDispatch::with_allowed_weight(weight_left);
+			let (queued_messages_dispatched, weight_spent_on_queued_dispatch) = queued_dispatcher.dispatch();
+			if queued_messages_dispatched != 0 {
+				frame_support::debug::trace!(
+					target: "runtime",
+					"Spent {} weight on {} queued messages from receive-messages",
+					weight_spent_on_queued_dispatch,
+					queued_messages_dispatched,
+				);
+			}
+
+			Ok(())
+		}
+
+		/// Receive messages receiving proof from bridged chain.
+		#[weight = 0] // TODO: update me (https://github.com/paritytech/parity-bridges-common/issues/78)
+		pub fn receive_message_receiving_proof(origin, proof: MessagesReceivingProofOf<T, I>) -> DispatchResult {
+			let _ = ensure_signed(origin)?;
+			let (lane_id, nonce) = T::TargetHeaderChain::verify_messages_receiving_proof(proof).map_err(|err| {
+				frame_support::debug::trace!(
+					target: "runtime",
+					"Rejecting invalid proof of messages receiving: {:?}",
+					err,
+				);
+
+				Error::<T, I>::InvalidMessagesReceivingProof
+			})?;
+
+			let mut lane = outbound_lane::<T, I>(lane_id);
+			let received_range = lane.confirm_receival(nonce);
+			if let Some(received_range) = received_range {
+				Self::deposit_event(Event::MessagesDelivered(lane_id, received_range.0, received_range.1));
+			}
+
+			frame_support::debug::trace!(
+				target: "runtime",
+				"Received proof of receiving messages up to (and including) {} at lane {:?}",
+				nonce,
+				lane_id,
+			);
+
+			Ok(())
+		}
+
+		/// Receive messages processing proof from bridged chain.
+		#[weight = 0] // TODO: update me (https://github.com/paritytech/parity-bridges-common/issues/78)
+		pub fn receive_message_processing_proof(origin, proof: MessagesProcessingProofOf<T, I>) -> DispatchResult {
+			let _ = ensure_signed(origin)?;
+			let (lane_id, nonce) = T::TargetHeaderChain::verify_messages_processing_proof(proof).map_err(|err| {
```

Comment:
Same as with delivery confirmation, this should give us the address of relayer who caused the message to be processed, so that we can reward him.
Since it's a `target` chain, we probably either need a map of `target -> source` addresses, or just have the relayers include their "reward" address of source chain, when they deliver / process messages in the target chain.

Reply:
I like the idea of the relayers being able to set a coinbase address to send their rewards to

Reply:
> Same as with delivery confirmation, this should give us the address of relayer who caused the message to be processed, so that we can reward him.

Responded in many other comments: (1) we should not reward redundant dispatch-transactions (2) rewards is not a part of this PR.

> Since it's a target chain, we probably either need a map of target -> source addresses, or just have the relayers include their "reward" address of source chain, when they deliver / process messages in the target chain.

Please see (II) of description/Some ideas for future PRs.